src/half_bridge_tim.cpp

```diff
@@ -0,0 +1,283 @@
+
+
+static bool _enabled;
+
+class PWM_TIM3
+{
+    public:
+
+    static void _init_registers(uint32_t pwm_resolution)
+    {
+        // Enable peripheral clock of GPIOB
+#if defined(STM32F0)
+        RCC->AHBENR |= RCC_AHBENR_GPIOBEN;
+#elif defined(STM32L0)
+        RCC->IOPENR |= RCC_IOPENR_IOPBEN;
+#endif
+
+        // Enable TIM3 clock
+        RCC->APB1ENR |= RCC_APB1ENR_TIM3EN;
+
+        // Select alternate function mode on PB0 and PB1 (first bit _1 = 1, second bit _0 = 0)
+#if defined(STM32F0)
+        GPIOB->MODER = (GPIOB->MODER & ~(GPIO_MODER_MODER0)) | GPIO_MODER_MODER0_1;
+        GPIOB->MODER = (GPIOB->MODER & ~(GPIO_MODER_MODER1)) | GPIO_MODER_MODER1_1;
+#elif defined(STM32L0)
+        GPIOB->MODER = (GPIOB->MODER & ~(GPIO_MODER_MODE0)) | GPIO_MODER_MODE0_1;
+        GPIOB->MODER = (GPIOB->MODER & ~(GPIO_MODER_MODE1)) | GPIO_MODER_MODE1_1;
+#endif
+
+        // Select AF2 on PB0 and PB1
+        GPIOB->AFR[0] |= 0x2 << GPIO_AFRL_AFRL0_Pos;
+        GPIOB->AFR[0] |= 0x2 << GPIO_AFRL_AFRL1_Pos;
+
+        // No prescaler --> timer frequency == SystemClock
+        TIM3->PSC = 0;
+
+        // Capture/Compare Mode Register 1
+        // OCxM = 110: Select PWM mode 1 on OCx
+        // OCxPE = 1:  Enable preload register on OCx (reset value)
+        TIM3->CCMR2 |= TIM_CCMR2_OC3M_2 | TIM_CCMR2_OC3M_1 | TIM_CCMR2_OC3PE;
+        TIM3->CCMR2 |= TIM_CCMR2_OC4M_2 | TIM_CCMR2_OC4M_1 | TIM_CCMR2_OC4PE;
+
+        // Capture/Compare Enable Register
+        // CCxP: Active high polarity on OCx (default = 0)
+        TIM3->CCER &= ~(TIM_CCER_CC3P); // PB0 / TIM3_CH3: high-side
+        TIM3->CCER |= TIM_CCER_CC4P;    // PB1 / TIM3_CH4: low-side
+
+        // Control Register 1
+        // TIM_CR1_CMS = 01: Select center-aligned mode 1
+        // TIM_CR1_CEN =  1: Counter enable
+        TIM3->CR1 |= TIM_CR1_CMS_0 | TIM_CR1_CEN;
+
+        // Control Register 2
+        // OIS1 = OIS1N = 0: Output Idle State is set to off
+        //TIM3->CR2 |= ;
+
+        // Force update generation (UG = 1)
+        TIM3->EGR |= TIM_EGR_UG;
+
+        // Auto Reload Register
+        // center-aligned mode --> resolution is double
+        TIM3->ARR = pwm_resolution;
+    }
+
+    static void start()
+    {
+        // Capture/Compare Enable Register
+        // CCxE = 1: Enable the output on OCx
+        // CCxP = 0: Active high polarity on OCx (default)
+        // CCxNE = 1: Enable the output on OC1N
+        // CCxNP = 0: Active high polarity on OC1N (default)
+        TIM3->CCER |= TIM_CCER_CC3E;
+        TIM3->CCER |= TIM_CCER_CC4E;
+    }
+
+    static void stop()
+    {
+        TIM3->CCER &= ~(TIM_CCER_CC3E);
+        TIM3->CCER &= ~(TIM_CCER_CC4E);
+    }
+
+    static int32_t get_ccr()
+    {
+        return TIM3->CCR3;
+    }
+
+    static void set_ccr(uint32_t val)
+    {
+        TIM3->CCR3 = val;                    // high-side
+        TIM3->CCR4 = val + _deadtime_clocks; // low-side
+    }
+};
+
+#if defined(STM32F0)
```

Comment:
Shouldn't we have that statement also above for STM3L0 before defining the classe PWM_TIM3?